File: python/239.py
import collections


class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        # maintain index of decreasing  possiable maximium values
        q = collections.deque()
        res = []
        for i, n in enumerate(nums):
            if q and q[0] <= i-k:
                q.popleft()
            while q and nums[q[-1]] < n:
                q.pop()
            q.append(i)
            res.append(nums[q[0]])
        return res[k-1:]

sol = Solution()
print(sol.maxSlidingWindow([1,3,-1,-3,5,3,6,7], 3))

File: cpp/239.cpp
#include <deque>
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
  vector<int> maxSlidingWindow(vector<int> &nums, int k) {
    if(k == 0) return {};
    deque<int> q;
    vector<int> v;
    for (int i = 0; i < nums.size(); i++) {
      if (!q.empty() && q.front() <= i - k)
        q.pop_front();
      while (!q.empty() && nums[q.back()] < nums[i])
        q.pop_back();
      q.push_back(i);
      v.push_back(nums[q.front()]);
    }
    return vector<int>(v.begin() + k - 1, v.end());
  }
};

int main(){
  vector<int> v{1,3,-1,-3,5,3,6,7};
  for(int n : Solution().maxSlidingWindow(v, 3))
    cout << n << " ";
  cout << endl;
}
